pkg/api/norman/store/cluster/cluster_store.go

```diff
@@ -306,14 +315,42 @@ func loadDataFromTemplate(clusterTemplateRevision *v3.ClusterTemplateRevision, c
 			return nil, httperror.WrapAPIError(err, httperror.ServerError, "Error processing clusterTemplate answers")
 		}
 		keyParts := strings.Split(question.Variable, ".")
-		values.PutValue(dataFromTemplate, val, keyParts...)
+		if strings.HasPrefix(question.Variable, "rancherKubernetesEngineConfig.privateRegistries") {
+			// for example: question.Variable = rancherKubernetesEngineConfig.privateRegistries[0].url
+			key := keyParts[len(keyParts)-1]
+			index, err := getIndexFromQuestion(question.Variable)
+			if err != nil {
+				return nil, httperror.WrapAPIError(err, httperror.ServerError, "Error processing clusterTemplate answers to private registry")
+			}
+			registry, ok := registryMap[index]
+			if ok {
+				registry[key] = val
+				registryMap[index] = registry
+			} else {
+				registryMap[index] = map[string]interface{}{
+					key: val,
+				}
+			}
+		} else {
+			values.PutValue(dataFromTemplate, val, keyParts...)
+		}
 
 		questionMap, err := convert.EncodeToMap(question)
 		if err != nil {
 			return nil, httperror.WrapAPIError(err, httperror.ServerError, "Error reading clusterTemplate questions")
 		}
 		revisionQuestions = append(revisionQuestions, questionMap)
 	}
+	registries := convertRegistryMapToSliceInOrder(registryMap)
+	// set the first registry in the list as the default registry
```

Comment:
Do we want to set it to `false` for the remaining ones?
Also, can `registries[0]` be `nil` if `registries` is a non-empty slice? I don't think it is possible. If there are existing registries, there will always be something in `registryMap` at key 0 which is then used to build `registries` and I don't see it ever removed from there. If it's empty to start, we would really be setting zero-index element via `privateRegistries[0]` (if we try to set something rather than `privateRegistries[0]` then `convertRegistryMapToSliceInOrder` will fail per another comment).
In this case we can simply check with registries are non-empty and set `registries[0]` instead of having for-loop.

Reply:
I updated the logic to set `isDefault` true for the 1st and false for the rest. 

I checked if registries[i] is nil because I was thinking about the case where the cluster template is created in a malformed format where the indexes for private registries are not continuous values (see my response in another comment).  But you are right that `convertRegistryMapToSliceInOrder` will fail first if that is the case. 

My latest implementation is based on the assumption that the indexes are continuous values( but the order could be random). This brings a general question: should we consider the above-mentioned special case?

Reply:
This is discussed offline and we decide to return an error if the index is not consecutive and less than the count of registries.